File: Desktop/agent-swarm-main/swarm/safety/loop_detector.py
from collections import defaultdict


class LoopDetector:
    def __init__(self, threshold: int = 3):
        self.threshold = threshold
        self._handoff_history: list[tuple[str, str]] = []
        self._pair_counts: dict[tuple[str, str], int] = defaultdict(int)

    def record_handoff(self, from_agent: str, to_agent: str):
        self._handoff_history.append((from_agent, to_agent))
        self._pair_counts[(from_agent, to_agent)] += 1

    def is_looping(self) -> bool:
        if len(self._handoff_history) < self.threshold:
            return False
        recent = self._handoff_history[-self.threshold:]
        if len(set(recent)) == 1:
            return True
        for pair, count in self._pair_counts.items():
            if count >= self.threshold:
                first_idx = next(
                    i for i, h in enumerate(self._handoff_history) if h == pair
                )
                if self._handoff_history[-1] == pair:
                    return True
        return False

    def detect_cycle(self) -> list:
        if len(self._handoff_history) < 4:
            return []
        recent = self._handoff_history[-6:]
        for length in range(2, len(recent) // 2 + 1):
            for start in range(len(recent) - length * 2 + 1):
                pattern = recent[start : start + length]
                if recent[start + length : start + length * 2] == pattern:
                    return pattern
        return []

    def summary(self) -> str:
        if not self._handoff_history:
            return "No handoffs yet"
        return " -> ".join(f"{a}->{b}" for a, b in self._handoff_history[-10:])

File: packages/opencode/src/swarm/loop-detector.ts
export class LoopDetector {
  private threshold: number
  private handoffHistory: Array<[string, string]>
  private pairCounts: Map<string, number>

  constructor(threshold: number = 3) {
    this.threshold = threshold
    this.handoffHistory = []
    this.pairCounts = new Map()
  }

  recordHandoff(from: string, to: string): void {
    this.handoffHistory.push([from, to])
    const key = `${from}->${to}`
    this.pairCounts.set(key, (this.pairCounts.get(key) ?? 0) + 1)
  }

  isLooping(): boolean {
    if (this.handoffHistory.length < this.threshold) return false

    const last = this.handoffHistory.slice(-this.threshold)
    const firstPair = `${last[0][0]}->${last[0][1]}`
    if (last.every(([f, t]) => `${f}->${t}` === firstPair)) return true

    const lastHandoff = this.handoffHistory[this.handoffHistory.length - 1]
    const lastKey = `${lastHandoff[0]}->${lastHandoff[1]}`
    const count = this.pairCounts.get(lastKey) ?? 0
    if (count >= this.threshold) return true

    return false
  }

  detectCycle(): string[] {
    if (this.handoffHistory.length < 4) return []

    const recent = this.handoffHistory.slice(-6)
    for (let patternLen = 2; patternLen <= Math.floor(recent.length / 2); patternLen++) {
      const pattern = recent.slice(-patternLen)
      let isRepeating = true
      for (let i = 0; i < patternLen && i + patternLen < recent.length; i++) {
        if (recent[i][0] !== recent[i + patternLen][0] || recent[i][1] !== recent[i + patternLen][1]) {
          isRepeating = false
          break
        }
      }
      if (isRepeating) return pattern.map(([f, t]) => `${f}->${t}`)
    }

    return []
  }

  summary(): string {
    const last = this.handoffHistory.slice(-10)
    if (last.length === 0) return "No handoffs yet"
    return last.map(([f, t]) => `${f}->${t}`).join(" -> ")
  }

  reset(): void {
    this.handoffHistory = []
    this.pairCounts.clear()
  }
}
